ACM#9987: HCP: Document Agent provisioning flow for non BM machines

xenolinux opened the pull request:
https://issues.redhat.com/browse/ACM-9987

@avishayt @CrystalChun @eranco74 @jparrill PTAL

xenolinux committed:
OSDOCS#9358: HCP: Document Agent provisioning flow for non BM machines

CrystalChun commented on clusters/hosted_control_planes/destroy_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,35 @@
+
++
+----
+oc delete managedcluster <cluster_name>
```
Quick Q: does a `managedcluster` always get created when running `hcp create cluster`?

CrystalChun replied:
And if it does, how come it's not automatically deleted when running `hcp destroy cluster`?

xenolinux replied:
TBH I am not sure. But I noticed that we use the `oc delete managedcluster <cluster_name>` command to destroy hosted clusters on [aws](https://github.com/stolostron/rhacm-docs/blob/2.10_stage/clusters/hosted_control_planes/destroy_cluster_aws.adoc), [OpenShift Virtualization](https://github.com/stolostron/rhacm-docs/blob/2.10_stage/clusters/hosted_control_planes/destroy_cluster_kubevirt.adoc), [bare metal](https://github.com/stolostron/rhacm-docs/blob/2.10_stage/clusters/hosted_control_planes/destroy_hosted_bm.adoc#destroying-a-hosted-cluster-on-bare-metal-by-using-the-command-line) as well. Should it just be this`hcp destroy cluster agent --name <cluster_name>` then? I can update other sections too if required

xenolinux replied:
I can update other sections too if required

jparrill replied:
We recommend to manage HostedClusters in the same way from creation to deletion. 

e.g: if you create the clusters using manifests, delete them using manifests with `oc apply -f <file>` and `oc delete -f <file>` andalso with `hcp` CLI.

If you use the `hcp` CLI to create the HostedCluster, a new finalizer is included in the HC, so if you creates the HC using the CLI and deletes it using `oc delete xxxx` it will get stuck in deletion.

xenolinux replied:
Ack'ed. Thanks for the confirmation. I removed the  `oc delete managedcluster <cluster_name>` command.

CrystalChun commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,126 @@
+hcp create cluster agent \
+    --name=${HOSTED_CLUSTER_NAME} \
+    --pull-secret=${PULL_SECRET_FILE} \
+    --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \
```
If a note could be added here to ensure this is the namespace where your `Agent` CRs exist (which can be found by `oc get agent -n <this namespace>`

CrystalChun replied:
Might also be nice to have a link or reference to how they can create Agents too

xenolinux replied:
@CrystalChun Do you have any resource link handy to provide it here? I couldn't find one

CrystalChun replied:
Yes here you go! https://access.redhat.com/documentation/en-us/red_hat_advanced_cluster_management_for_kubernetes/2.9/html/clusters/cluster_mce_overview#add-host-host-inventory 
Will defer to you and your team for where to best link this!

CrystalChun commented on clusters/hosted_control_planes/non_bm_intro.adoc:
```diff
@@ -0,0 +1,82 @@
+
+- When you scale up a node pool, a machine is created. The Cluster API provider finds an Agent that is approved, is passing validations, is not currently in use, and meets the requirements that are specified in the node pool specification. You can monitor the installation of an Agent by checking its status and conditions.
+
+- When you scale down a node pool, Agents are unbound from the corresponding cluster. Before you can reuse the clusters, you must restart them by using the Discovery image to update the number of nodes.
```
```suggestion
- When you scale down a node pool, Agents are unbound from the corresponding cluster. Before you can reuse the Agents, you must restart them by using the Discovery image.
```

swopebe replied:
This was merged but I was in the middle of a peer review, will reach out over slack. I should have made it clear that I was reviewing still...sorry.

CrystalChun commented on clusters/hosted_control_planes/non_bm_intro.adoc:
```diff
@@ -0,0 +1,82 @@
+
+- You can add agent machines as a worker node to a hosted cluster by using the Agent platform. Agent machine represents a host booted with a Discovery Image and ready to be provisioned as an {ocp-short} node. The Agent platform is part of the central infrastructure management service. For more information, see xref:../cluster_lifecycle/cim_enable.adoc#enable-cim[Enabling the central infrastructur ...
+
+- You can boot your non bare metal hosts with the ISO only.
+
+- Each host must be started manually with a Discovery Image that the central infrastructure management provides.
```
```suggestion
- Each non bare metal host must be booted manually with a Discovery Image ISO that the central infrastructure management provides.
```

swopebe replied:
non bare metal probably should not be used, IMO

RECOMMEND: All hosts that are not bare metal require a manual boot with a...

xenolinux replied:
Addressed

swopebe replied:
I see the file name is even non bm, so we may need to keep it. We are trying to shorten filenames in hosted cluster, though.

CrystalChun commented on clusters/hosted_control_planes/non_bm_intro.adoc:
```diff
@@ -0,0 +1,82 @@
+
+- When you create a hosted cluster with the Agent platform, HyperShift installs the Agent Cluster API provider in the hosted control plane namespace.
+
+- When you scale up a node pool, a machine is created. The Cluster API provider finds an Agent that is approved, is passing validations, is not currently in use, and meets the requirements that are specified in the node pool specification. You can monitor the installation of an Agent by checking its status and conditions.
```
```suggestion
- A machine is created for every replica you scale the node pool up by. For every machine, the Cluster API provider finds and installs an Agent that is approved, is passing validations, is not currently in use, and meets the requirements that are specified in the node pool specification. You can monitor the installation of an Agent by checking its status and conditions.
```

swopebe replied:
@CrystalChun we have to try to write active voice and not passive voice. That is why the first line is preferred. Is the original line incorrect somehow? 

When you scale up a node pool, a machine is created.

Can say _Each time you scale..._

xenolinux replied:
We have the first sentence[1] already present in the bm_intro.adoc file as well. So IMO, the sentence must not be inaccurate. 
Hence, I changed it to : 
When you scale up the node pool, a machine is created for every replica.

jparrill commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,112 @@
+
++
+----
+oc --kubeconfig ${HOSTED_CLUSTER_NAME}.kubeconfig get nodes
```
I don't know if this was mentioned before, but you will need to extract the `admin-kubeconfig` secret of the HostedCluster previous to these commands executions. If that was mentioned before, perfect :), if not, this is the command:

```
oc extract -n clusters secret/hosted-admin-kubeconfig --to=./hostedcluster-secrets --confirm
```

output:
```
hostedcluster-secrets/kubeconfig
```

jparrill commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,127 @@
+export PULL_SECRET_FILE=$PWD/pull-secret
+export MACHINE_CIDR=192.168.122.0/24
+export ETCD_STORAGE="lvm-storageclass"
+export OCP_RELEASE=4.14.0-multi
```
I would change this by an official stable 4.14 version (e.g `4.14.13-x86_64`)

swopebe replied:
How we are writing these, to avoid creating bugs when we can, is with a variable:

`OCP_RELEASE=4.x.0-multi`

With either a call out or a simple line that says, "Replace 4.x.0 with the supported version" (or something like that, there are examples in the doc that we recently changed).

jparrill commented on clusters/hosted_control_planes/handle_ingress_non_bm.adoc:
```diff
@@ -0,0 +1,213 @@
+    - ${INGRESS_IP}-${INGRESS_IP}
+---
+apiVersion: metallb.io/v1beta1
+kind: BGPAdvertisement
```
The BGPAdvertisement works for you? In my env I've only make the L2Advertisement work properly.

xenolinux replied:
sry, I haven't tried `BGPAdvertisement`. This module is copied from the [pre-existing bare metal section](https://github.com/stolostron/rhacm-docs/blob/2.10_stage/clusters/hosted_control_planes/handle_ingress_bm.adoc)

jparrill commented on clusters/hosted_control_planes/handle_ingress_non_bm.adoc:
```diff
@@ -0,0 +1,213 @@
++
+----
+NAME                                         VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
+clusterversion.config.openshift.io/version   4.12z    True        False         3m32s   Cluster version is 4.12z
```
I would change the 4.12z to the version you mentioned at the beginning of the PR. Have in mind to do that also in the nexts ones.

xenolinux replied:
Updated `4.12.z` to `4.14.13-x86_64`

swopebe replied:
Same here, we are using variables when we can.

jparrill commented on clusters/hosted_control_planes/non_bm_intro.adoc:
```diff
@@ -0,0 +1,80 @@
+
+- You can add agent machines as a worker node to a hosted cluster by using the Agent platform. Agent machine represents a host booted with a Discovery Image and ready to be provisioned as an {ocp-short} node. The Agent platform is part of the central infrastructure management service. For more information, see xref:../cluster_lifecycle/cim_enable.adoc#enable-cim[Enabling the central infrastructur ...
+
+- Each non bare metal host must be booted manually with a Discovery Image ISO that the central infrastructure management provides.
```
put emphasis that this boot should be bootOnce, if not the customer will stay in a loop when the workers get rebooted by the first time.

xenolinux replied:
Is the `bootOnce` a value that can be set somewhere while booting? Or just mentioning "Each non bare metal host must be booted manually only once with a Discovery Image....." is enough?

jparrill replied:
No, it's just a way to boot a node. Is enough just to mention it, the bootOnce procedure is just a helper for the user but a parameter or flag in any place. This will depend in the customers BMC, Virtualiser, etc.. so just mentioning that should be enough

sample:

"Make sure you mount the LiveISO in the worker node in BootOnce mode. If not you will need to unmount that LiveISO from the worker once the first reboot happens."

jparrill reviewed (changes requested):
Looking good :). Dropped some comments.

jparrill commented:
/lgtm

CrystalChun commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,127 @@
+----
+
++
+If 10 minutes pass without requiring the additional capacity, the worker node is removed. On the Agent platform, the agent is decommissioned and can be reused.
```
I wonder if this line should be moved down here https://github.com/stolostron/rhacm-docs/pull/6073/files#diff-241549552c1d12c34d4987d8dfbf5c0c48a17e05fb2854a527bd2464cb3f61bdR101 wdyt?

CrystalChun commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,127 @@
+oc -n ${CLUSTERS_NAMESPACE} patch nodepool ${HOSTED_CLUSTER_NAME} --type=json -p '[\{"op":"remove", "path": "/spec/autoScaling"}, \{"op": "add", "path": "/spec/replicas", "value": $SOME_INT_VALUE_FOR_SCALING_TO}]'
+----
+
+The command removes `"spec.autoScaling"`` from the YAML file, adds `"spec.replicas"`, and sets `"spec.replicas"` to the value that you specify.
```
```suggestion
The command removes `"spec.autoScaling"` from the YAML file, adds `"spec.replicas"`, and sets `"spec.replicas"` to the value that you specify.
```

CrystalChun commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,127 @@
+----
+
++
+<1> Typically, the namespace is created by the HyperShift Operator, but agent cluster creation generates a cluster API provider role that needs the namespace to already exist.
+<2> Ensure that agents are available in the `${HOSTED_CONTROL_PLANE_NAMESPACE}`. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.
```
Sorry, I realize now that having a note in the command the user would run is not such a great idea 😅 Suggestion to merge it into <1> but feel free to do what you think is best!
```suggestion
<1> Typically, the namespace is created by the HyperShift Operator, but agent cluster creation generates a cluster API provider role that needs the namespace to already exist. In the command below, it is also used as the `agent-namespace` flag. Ensure that agents are available in the `${HOSTED_CONTROL_PLANE_NAMESPACE}`. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.
```

CrystalChun replied:
Checking if this change was meant to occur, if it wasn't, feel free to ignore this comment!

CrystalChun commented on clusters/hosted_control_planes/scale_nodepool_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,175 @@
+BMH: ocp-worker-1 Agent: da503cf1-a347-44f2-875c-4960ddb04091 State: insufficient
+----
+
+. After the agents reach the `added-to-existing-cluster` state, verify that you can see the {ocp-short} nodes by entering the following command:
```
```suggestion
. After the agents reach the `added-to-existing-cluster` state, verify that you can see the {ocp-short} nodes in the hosted cluster by entering the following command:
```

Or maybe mentioning to extract the hosted cluster kubeconfig prior to this?

CrystalChun commented on clusters/hosted_control_planes/scale_nodepool_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,175 @@
+
+You can create node pools for a hosted cluster by specifying a name, number of replicas, and any additional information, such as an agent label selector.
+
+. To create a node pool, enter the following information. In this example, the node pool uses agents with the `"size" : "medium"` label:
```
Could we mention this label selector flag is optional?

CrystalChun commented on clusters/hosted_control_planes/scale_nodepool_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,175 @@
+
++
+----
+oc --kubeconfig $CLUSTER_NAME-kubeconfig get nodes
```
Similar to https://github.com/stolostron/rhacm-docs/pull/6073/files#r1492723187
maybe mentioning to extract the hosted cluster kubeconfig

CrystalChun commented on clusters/hosted_control_planes/scale_nodepool_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,175 @@
+oc --kubeconfig $CLUSTER_NAME-kubeconfig get nodes
+----
+
+. Verify that the node pool is in the status that you expect by entering this command:
```
Maybe adding expected status output or stating that the number of available nodes matches the number of expected nodes?

CrystalChun commented on clusters/hosted_control_planes/verify_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,138 @@
+
++
+----
+oc extract -n kni21 secret/kni21-admin-kubeconfig --to=- > kubeconfig-kni21
```
Maybe for consistency with the rest of the docs to let the user know which fields to replace such as:

```suggestion
oc extract -n <hosted cluster namespace> secret/<hosted cluster name>-admin-kubeconfig --to=- > kubeconfig-<hosted cluster name>
```

CrystalChun commented on clusters/hosted_control_planes/verify_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,138 @@
+
++
+----
+oc get co --kubeconfig=kubeconfig-kni21
```
```suggestion
oc get co --kubeconfig=kubeconfig-<hosted cluster name>
```

CrystalChun commented on clusters/hosted_control_planes/verify_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,138 @@
+
++
+----
+oc get pods -A --kubeconfig=kubeconfig-kni21
```
```suggestion
oc get pods -A --kubeconfig=kubeconfig-<hosted cluster name>
```

avishayt commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,124 @@
+oc --kubeconfig ${HOSTED_CLUSTER_NAME}.kubeconfig -n default delete deployment reversewords
+----
+
+. If 10 minutes pass without requiring the additional capacity, the worker node is removed. On the Agent platform, the agent is decommissioned and can be reused. You can confirm that the node was removed by entering the following command:
```
I don't think we should promise exactly 10 minutes.  More like "Once the system has seen that capacity is not needed for a long enough time..."

xenolinux replied:
Addressed by rewording the sentence a bit.

avishayt commented:
Overall, it looks OK.  Are we sure that we want to copy-paste all of these sections?  I'm worried that we'll update one copy and not the other, leading to diverging instructions that are 99.9% the same.

xenolinux commented:
> Overall, it looks OK. Are we sure that we want to copy-paste all of these sections?
@avishayt The copy-paste looks like a feasible option here considering this repo does not use assembly files and we need a separate section for non bare metal use case. 

> I'm worried that we'll update one copy and not the other, leading to diverging instructions that are 99.9% the same.
You mean the bare metal copy and non bare metal copy being different from one another?

avishayt commented:
> You mean the bare metal copy and non bare metal copy being different from one another?

Yes, we need to make sure to always update both

xenolinux commented:
> > You mean the bare metal copy and non bare metal copy being different from one another?
> 
> Yes, we need to make sure to always update both

Yes I have plans to update that section as well. Once reviews are done, I will raise a PR for updating bare metal sections.

CrystalChun commented on clusters/hosted_control_planes/scale_nodepool_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,200 @@
+
++
+----
+oc extract -n <hosted-cluster-namespace> secret/<hosted-cluster-name>-admin-kubeconfig --to=- > kubeconfig-<hosted-cluster-name>
+----
+
+. After the agents reach the `added-to-existing-cluster` state, verify that you can see the {ocp-short} nodes in the hosted cluster by entering the following command:
+
++
+----
+oc --kubeconfig <hosted-cluster-name>.kubeconfig get nodes
+----
```
Minor: let's match the name kubeconfig file name to the filename of the one that's extracted? 
```suggestion
oc extract -n <hosted-cluster-namespace> secret/<hosted-cluster-name>-admin-kubeconfig --to=- > kubeconfig-<hosted-cluster-name>
----

. After the agents reach the `added-to-existing-cluster` state, verify that you can see the {ocp-short} nodes in the hosted cluster by entering the following command:

+
----
oc --kubeconfig kubeconfig-<hosted-cluster-name> get nodes
----
```

CrystalChun commented on clusters/hosted_control_planes/scale_nodepool_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,200 @@
+
++
+----
+oc --kubeconfig <hosted-cluster-name>.kubeconfig get clusterversion,co
```
```suggestion
oc --kubeconfig kubeconfig-<hosted-cluster-name> get clusterversion,co
```

CrystalChun commented on clusters/hosted_control_planes/scale_nodepool_hosted_non_bm.adoc:
```diff
@@ -0,0 +1,200 @@
+
++
+----
+oc --kubeconfig $CLUSTER_NAME-kubeconfig get nodes
```
Similar, match name of kubeconfig file to the one that is extracted.
```suggestion
oc --kubeconfig ./hostedcluster-secrets get nodes
```

CrystalChun commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,128 @@
+
++
+----
+export CLUSTERS_NAMESPACE="clusters"
+export HOSTED_CLUSTER_NAME="example"
+export HOSTED_CONTROL_PLANE_NAMESPACE="${CLUSTERS_NAMESPACE}-${HOSTED_CLUSTER_NAME}"
+export BASEDOMAIN="krnl.es"
+export PULL_SECRET_FILE=$PWD/pull-secret
+export MACHINE_CIDR=192.168.122.0/24
+export ETCD_STORAGE="lvm-storageclass"
+export OCP_RELEASE=4.14.13-x86_64
+export SSH_KEY=${HOME}/.ssh/id_rsa.pub
+oc create ns ${HOSTED_CONTROL_PLANE_NAMESPACE}
+
+hcp create cluster agent \
+    --name=${HOSTED_CLUSTER_NAME} \
+    --pull-secret=${PULL_SECRET_FILE} \
+    --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \ <1>
+    --base-domain=${BASEDOMAIN} \
+    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN} \
+    --etcd-storage-class=${ETCD_STORAGE} \
+    --ssh-key  ${SSH_KEY} \
+    --namespace ${CLUSTERS_NAMESPACE} \
+    --control-plane-availability-policy SingleReplica \
+    --release-image=quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE} \
+    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN}
+----
+
++
+<1> Required by the agent cluster API provider and must already exist. You can use the `--agent-namespace` flag in the `hcp create cluster` command. Ensure that agents are available in this namespace. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.
```
```suggestion
export CLUSTERS_NAMESPACE="clusters"
export HOSTED_CLUSTER_NAME="example"
export HOSTED_CONTROL_PLANE_NAMESPACE="${CLUSTERS_NAMESPACE}-${HOSTED_CLUSTER_NAME}" <1>
export BASEDOMAIN="krnl.es"
export PULL_SECRET_FILE=$PWD/pull-secret
export MACHINE_CIDR=192.168.122.0/24
export ETCD_STORAGE="lvm-storageclass"
export OCP_RELEASE=4.14.13-x86_64
export SSH_KEY=${HOME}/.ssh/id_rsa.pub
oc create ns ${HOSTED_CONTROL_PLANE_NAMESPACE}

hcp create cluster agent \
    --name=${HOSTED_CLUSTER_NAME} \
    --pull-secret=${PULL_SECRET_FILE} \
    --agent-namespace=${HOSTED_CONTROL_PLANE_NAMESPACE} \
    --base-domain=${BASEDOMAIN} \
    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN} \
    --etcd-storage-class=${ETCD_STORAGE} \
    --ssh-key  ${SSH_KEY} \
    --namespace ${CLUSTERS_NAMESPACE} \
    --control-plane-availability-policy SingleReplica \
    --release-image=quay.io/openshift-release-dev/ocp-release:${OCP_RELEASE} \
    --api-server-address=api.${HOSTED_CLUSTER_NAME}.${BASEDOMAIN}
----

+
<1> Required by the agent cluster API provider and must already exist. Use this as the `--agent-namespace` flag in the `hcp create cluster` command. Ensure that agents are available in this namespace. You can find the list of agents by using the `oc get agent -n ${HOSTED_CONTROL_PLANE_NAMESPACE}` command.
```

CrystalChun reviewed (approved):
/lgtm

oafischer commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,124 @@
+
+When you need more capacity in your hosted cluster and spare agents are available, you can enable auto-scaling to install new worker nodes.
+
+. To enable auto-scaling, enter the following command. In this case, the minimum number of nodes is 2, and the maximum is 5. The maximum number of nodes that you can add might be bound by your platform. For example, if you use the Agent platform, the maximum number of nodes is bound by the number of available agents.
```
```suggestion
. To enable auto-scaling, enter the following command. In the example, the minimum number of nodes is 2, and the maximum is 5. The maximum number of nodes that you can add might be bound by your platform. For example, if you use the Agent platform, the maximum number of nodes is bound by the number of available agents:
```

oafischer commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,124 @@
+
+. Create a workload that requires a new node.
+
+.. Create a YAML file that contains the workload configuration, as shown in the following example:
```
You can probably avoid passive voice here if you replace "as shown" with "by using"

oafischer commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,124 @@
+oc --kubeconfig ${HOSTED_CLUSTER_NAME}.kubeconfig get nodes
+----
+
++
+See the following example output:
+
++
+----
+NAME           STATUS   ROLES    AGE   VERSION
+ocp-worker-0   Ready    worker   35s   v1.24.0+3882f8f
+ocp-worker-1   Ready    worker   40m   v1.24.0+3882f8f
+ocp-worker-2   Ready    worker   41m   v1.24.0+3882f8f
+----
```
We can shorten this by just stating what the use should look for if they want to verify success. I assume we are looking for the `Ready` status of the worker nodes, so I don't think we need this output table.

oafischer commented on clusters/hosted_control_planes/autoscale_node_hosted_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,124 @@
+----
+
++
+See the following example output:
+
++
+----
+NAME           STATUS   ROLES    AGE   VERSION
+ocp-worker-1   Ready    worker   51m   v1.24.0+3882f8f
+ocp-worker-2   Ready    worker   52m   v1.24.0+3882f8f
+----
```
Same here, we usually try to be sparing with example outputs

oafischer commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,128 @@
+oc -n ${HOSTED_CONTROL_PLANE_NAMESPACE} get pods
+----
+
++
+See the following example output:
+
++
+----
+NAME                                             READY   STATUS    RESTARTS   AGE
+capi-provider-7dcf5fc4c4-nr9sq                   1/1     Running   0          4m32s
+catalog-operator-6cd867cc7-phb2q                 2/2     Running   0          2m50s
+certified-operators-catalog-884c756c4-zdt64      1/1     Running   0          2m51s
+cluster-api-f75d86f8c-56wfz                      1/1     Running   0          4m32s
+cluster-autoscaler-7977864686-2rz4c              1/1     Running   0          4m13s
+cluster-network-operator-754cf4ffd6-lwfm2        1/1     Running   0          2m51s
+cluster-policy-controller-784f995d5-7cbrz        1/1     Running   0          2m51s
+cluster-version-operator-5c68f7f4f8-lqzcm        1/1     Running   0          2m51s
+community-operators-catalog-58599d96cd-vpj2v     1/1     Running   0          2m51s
+control-plane-operator-f6b4c8465-4k5dh           1/1     Running   0          4m32s
+etcd-0                                           1/1     Running   0          4m13s
+hosted-cluster-config-operator-c4776f89f-dt46j   1/1     Running   0          2m51s
+ignition-server-7cd8676fc5-hjx29                 1/1     Running   0          4m22s
+ingress-operator-75484cdc8c-zhdz5                1/2     Running   0          2m51s
+konnectivity-agent-c5485c9df-jsm9s               1/1     Running   0          4m13s
+konnectivity-server-85dc754888-7z8vm             1/1     Running   0          4m13s
+kube-apiserver-db5fb5549-zlvpq                   3/3     Running   0          4m13s
+kube-controller-manager-5fbf7b7b7b-mrtjj         1/1     Running   0          90s
+kube-scheduler-776c59d757-kfhv6                  1/1     Running   0          3m12s
+machine-approver-c6b947895-lkdbk                 1/1     Running   0          4m13s
+oauth-openshift-787b87cff6-trvd6                 2/2     Running   0          87s
+olm-operator-69c4657864-hxwzk                    2/2     Running   0          2m50s
+openshift-apiserver-67f9d9c5c7-c9bmv             2/2     Running   0          89s
+openshift-controller-manager-5899fc8778-q89xh    1/1     Running   0          2m51s
+openshift-oauth-apiserver-569c78c4d-568v8        1/1     Running   0          2m52s
+packageserver-ddfffb8d7-wlz6l                    2/2     Running   0          2m50s
+redhat-marketplace-catalog-7dd77d896-jtxkd       1/1     Running   0          2m51s
+redhat-operators-catalog-d66b5c965-qwhn7         1/1     Running   0          2m51s
+----
```
This is definitely an output that is too long and can be replaced by a sentence telling the user to check the `STATUS` of their pods, if they need verification.

oafischer commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,128 @@
+
+. In the console header, ensure that **All Clusters** is selected.
+
+. Click **Infrastructure -> Clusters**.
```
```suggestion
. Click **Infrastructure > Clusters**.
```

We just use a simple `>` for these, make sure to replace them in the rest of the PR as well.

oafischer commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,128 @@
+----
+
+[#hosted-create-non-bare-metal-console]
+== Creating a hosted cluster on non bare metal agent machines by using the console
```
We have some topics that document console processes, but we generally try to avoid documenting the console. I'm not sure if this would be considered too much, what do you think @swopebe?

swopebe replied:
Yes, we don't write UI steps unless 1. We have to document how to use a third-party console (Grafana) or 2. The console is not intuitive.

I think by using "follow the prompts" that this may be as concise as it can be, not sure the user could go to "create cluster" without these bold steps. 

I see some changes that need to be made, so I will review this in the afternoon, as well.

xenolinux replied:
IMO, since we cover creating hosted clusters using console for bare metal, the console procedure would make sense here for non bare metal as well.

xenolinux replied:
> Yes, we don't write UI steps unless 1. We have to document how to use a third-party console (Grafana) or 2. The console is not intuitive.
> 
> I think by using "follow the prompts" that this may be as concise as it can be, not sure the user could go to "create cluster" without these bold steps.
> 
> I see some changes that need to be made, so I will review this in the afternoon, as well.

Sure thanks

oafischer commented on clusters/hosted_control_planes/handle_ingress_non_bm.adoc:
```diff
@@ -0,0 +1,213 @@
+[#handling-ingress-non-bm]
+= Handling ingress in a hosted cluster on non bare metal agent machines
+
+Every {ocp-short} cluster comes set up with a default application Ingress Controller that is expected have an external DNS record associated with it. For example, if you create a HyperShift cluster named `example` with the base domain `krnl.es`, you can expect the wildcard domain
```
```suggestion
Every {ocp-short} cluster has a default application Ingress Controller that is expected to have an external DNS record associated with it. For example, if you create a HyperShift cluster named `example` with the base domain `krnl.es`, you can expect the wildcard domain
```

oafischer commented on clusters/hosted_control_planes/handle_ingress_non_bm.adoc:
```diff
@@ -0,0 +1,213 @@
+Every {ocp-short} cluster comes set up with a default application Ingress Controller that is expected have an external DNS record associated with it. For example, if you create a HyperShift cluster named `example` with the base domain `krnl.es`, you can expect the wildcard domain
+`*.apps.example.krnl.es` to be routable.
+
+You can set up a load balancer and wildcard DNS record for the `*.apps`. This process requires deploying MetalLB, configuring a new load balancer service that routes to the ingress deployment, and assigning a wildcard DNS entry to the load balancer IP address.
```
```suggestion
You can set up a load balancer and wildcard DNS record for the `*.apps`. The process requires deploying MetalLB, configuring a new load balancer service that routes to the ingress deployment, and assigning a wildcard DNS entry to the load balancer IP address. Complete the following steps:
```

oafischer commented on clusters/hosted_control_planes/handle_ingress_non_bm.adoc:
```diff
@@ -0,0 +1,213 @@
+----
+
++
+See the following example output:
+
++
+----
+NAME                                         VERSION           AVAILABLE   PROGRESSING   SINCE   STATUS
+clusterversion.config.openshift.io/version   4.14.13-x86_64    True        False         3m32s   Cluster version is 4.14.13-x86_64
+
+NAME                                                                           VERSION           AVAILABLE   PROGRESSING   DEGRADED   SINCE   MESSAGE
+clusteroperator.config.openshift.io/console                                    4.14.13-x86_64    True        False         False      3m50s
+clusteroperator.config.openshift.io/csi-snapshot-controller                    4.14.13-x86_64    True        False         False      25m
+clusteroperator.config.openshift.io/dns                                        4.14.13-x86_64    True        False         False      23m
+clusteroperator.config.openshift.io/image-registry                             4.14.13-x86_64    True        False         False      23m
+clusteroperator.config.openshift.io/ingress                                    4.14.13-x86_64    True        False         False      53m
+clusteroperator.config.openshift.io/insights                                   4.14.13-x86_64    True        False         False      25m
+clusteroperator.config.openshift.io/kube-apiserver                             4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/kube-controller-manager                    4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/kube-scheduler                             4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/kube-storage-version-migrator              4.14.13-x86_64    True        False         False      25m
+clusteroperator.config.openshift.io/monitoring                                 4.14.13-x86_64    True        False         False      21m
+clusteroperator.config.openshift.io/network                                    4.14.13-x86_64    True        False         False      25m
+clusteroperator.config.openshift.io/openshift-apiserver                        4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/openshift-controller-manager               4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/openshift-samples                          4.14.13-x86_64    True        False         False      23m
+clusteroperator.config.openshift.io/operator-lifecycle-manager                 4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/operator-lifecycle-manager-catalog         4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/operator-lifecycle-manager-packageserver   4.14.13-x86_64    True        False         False      54m
+clusteroperator.config.openshift.io/service-ca                                 4.14.13-x86_64    True        False         False      25m
+clusteroperator.config.openshift.io/storage                                    4.14.13-x86_64    True        False         False      25m
+----
```
Not sure what we are looking for here but we need to shorten/remove this output as well and let the user know what they need to look for.

oafischer commented on clusters/hosted_control_planes/non_bm_intro.adoc:
```diff
@@ -0,0 +1,80 @@
+[#configuring-hosting-service-cluster-configure-agent-non-bm]
+= Configuring hosted control plane clusters using non bare metal agent machines (Tech Preview)
```
New tech preview items should also have an entry in the what's new file in the release notes section.

xenolinux replied:
Done https://github.com/stolostron/rhacm-docs/pull/6154

swopebe commented on clusters/hosted_control_planes/create_cluster_non_bm.adoc:
```diff
@@ -0,0 +1,104 @@
+
+. Monitor the deployment of the hosted cluster in the **Hosted cluster** view. If you do not see information about the hosted cluster, ensure that **All Clusters** is selected, and click the cluster name.
+
+. Wait until the control plane components are ready. This process can take a few minutes.
```
Wait is not really a step, so that should be with the other steps. I will play with it.

xenolinux commented:
Thank you for the review. I addressed all the suggestions and comments -- I will proceed with merging this PR. This content is intended to go in the OpenShift docs 4.15 welcome page (OCP 4.15 docs merge freeze date: 22nd Feb 2024)

If there are any minor comments/suggestions from the peer review perspective, I will make sure to open a follow up PR to address those changes.

oafischer commented:
The example outputs still need to be removed, we definitely can't include the long ones. The shorter ones can remain for now to keep things simple.
